table/block_based/block_based_table_builder.cc

```diff
@@ -1101,26 +1333,56 @@ void BlockBasedTableBuilder::EnterUnbuffered() {
                 r->compression_type == kZSTDNotFinalCompression));
 
   for (size_t i = 0; ok() && i < r->data_block_and_keys_buffers.size(); ++i) {
-    const auto& data_block = r->data_block_and_keys_buffers[i].first;
+    auto& data_block = r->data_block_and_keys_buffers[i].first;
     auto& keys = r->data_block_and_keys_buffers[i].second;
     assert(!data_block.empty());
     assert(!keys.empty());
 
-    for (const auto& key : keys) {
-      if (r->filter_builder != nullptr) {
-        size_t ts_sz =
-            r->internal_comparator.user_comparator()->timestamp_size();
-        r->filter_builder->Add(ExtractUserKeyAndStripTimestamp(key, ts_sz));
+    if (r->compression_opts.parallel_threads > 1) {
+      ParallelCompressionRep::BlockRep block_rep;
+      r->pc_rep->block_rep_pool.pop(block_rep);
+
+      std::swap(*(block_rep.data_ptr), data_block);
+      block_rep.contents = *(block_rep.data_ptr);
+
+      block_rep.compression_type = r->compression_type;
+
+      block_rep.keys_ptr->SwapAssign(keys);
+
+      if (i + 1 < r->data_block_and_keys_buffers.size()) {
+        block_rep.first_key_in_next_block_ptr->assign(
+            r->data_block_and_keys_buffers[i + 1].second.front());
+      } else {
+        if (r->first_key_in_next_block_ptr == nullptr) {
+          delete block_rep.first_key_in_next_block_ptr;
+          block_rep.first_key_in_next_block_ptr = nullptr;
+        } else {
+          block_rep.first_key_in_next_block_ptr->assign(
+              r->first_key_in_next_block_ptr->data(),
+              r->first_key_in_next_block_ptr->size());
+        }
+      }
+
+      r->pc_rep->write_queue.push(std::move(block_rep.slot_ptr));
+      r->pc_rep->compress_queue.push(std::move(block_rep));
```

Comment:
This `std::move()` is confusing to me too. I'm not sure about the behavior we want for `block_rep` after the move. 
My understanding is that, we want to reuse those allocated memory for keys, strings, first_key_in_next_block_ptr, etc. If that is the case, can we be more explicit here? If block_rep_pool keeps holding the ownership to all those objects pointed by those pointers, I don't think we should do std::move() here.

Either way, please add comments somewhere to explain the ownership for those objects.

Reply:
When block_rep is a struct, std::move will only move its value (including pointers), but not the data it references. Now block_rep is a pointer, std::move will only move its value as well. I've added comments for object ownership around BlockRep definition and WorkQueue variable definitions.

Reply:
Moving a pointer is confusing too. My vote would be to move away from this move.

Reply:
Modified WorkQueue design so it copies elements instead of moves them. As long as we avoid passing large elements directly to WorkQueue in future, this should be fine. We can always pass large objects by pointers.
This behavior is more similar to STL queue, and maybe less misleading.